Mathlib/Algebra/Group/Submonoid/Membership.lean

```diff
@@ -233,7 +233,7 @@ then it holds for all elements of the supremum of `S`. -/
 theorem iSup_induction {ι : Sort*} (S : ι → Submonoid M) {C : M → Prop} {x : M} (hx : x ∈ ⨆ i, S i)
     (mem : ∀ (i), ∀ x ∈ S i, C x) (one : C 1) (mul : ∀ x y, C x → C y → C (x * y)) : C x := by
   rw [iSup_eq_closure] at hx
-  refine closure_induction hx (fun x hx => ?_) one mul
+  refine closure_induction (fun x hx => ?_) one (fun _ _ _ _ ↦ mul _ _) hx
```

Comment:
It's the proliferation of underscores in things like this that I was also worried about - we go from writing `mul` as the term to writing `(fun _ _ _ _ ↦ mul _ _)` instead. Like I say, I don't think this is a huge problem, but it can make these induction lemmas more awkward to use in practice: here and in the proof of `closureCommMonoidOfComm` (which is the next thing in the git diff in this file), it's quite a bit less clear what's going on. In fact, for the not-particularly-similar case of finite sums, I wanted to introduce a non-dependent version for exactly this reason (Finset.sum_nbij), which turned out to be really practical. 

Again though, I won't block on this, the change in this PR is clearly overall positive.

Reply:
I think the idea is that you probably should be using `induction _ using closure_induction` instead of `refine closure_induction ..` in which case the underscore proliferation is less bad; but @j-loreaux deliberately did not make that change in this PR to reduce the overall diff.

Reply:
Indeed, for the most part I didn't switch to `induction _ using closure_induction` in order to keep the diff small (Although in the case of `closureCommMonoidOfComm`, it's just that it was low in the hierarchy and I didn't realize yet that those should be switched for readability).

As for `iSup_induction`, part of the problem is that it's *not* dependent, when it should be. Note that there is `iSup_induction'` which *is* dependent. I didn't do the replacement for those lemmas as well because I was trying to keep the changes pared down to the `closure/adjoin/span_induction` lemmas.

Here is a proof of the dependent version of `iSup_induction`. Note that using `key` to deal with the motive is essential, but this is not surprising because `⨆ i, S i` isn't defeq to `closure s` for some `s`. That's essentially why `iSup_induction` is it's own lemma in the first place.
```lean
theorem iSup_induction {ι : Sort*} (S : ι → Submonoid M) {C : (x : M) → (x ∈ ⨆ i, S i) → Prop}
    (mem : ∀ (i) (x) (hx : x ∈ S i), C x (mem_iSup_of_mem i hx)) (one : C 1 (one_mem _))
    (mul : ∀ x hx y hy, C x hx → C y hy → C (x * y) (mul_mem hx hy))
    {x : M} (hx : x ∈ ⨆ i, S i) : C x hx := by
  have key {y} : y ∈ ⨆ i, S i ↔ y ∈ closure (⋃ i, (S i : Set M)) := by simp only [iSup_eq_closure]
  show C x (key.mpr (key.mp hx))
  refine closure_induction ?_ ?_ ?_ (key.mp hx)
  all_goals simpa [← key, forall_comm (β := ι)]
```

Reply:
Indeed, `iSup_induction` isn't dependent, but that's exactly the situation I'm worried about: when you need to prove something which is not dependent, but all you have access to is a dependent induction

Reply:
But why do we want the non-dependent versions at all? They don't `elab_as_elim` properly because Lean can't compute the motive.

Reply:
Oh, for applications where you want to prove something that's not an induction lemma, you just want to use the induction lemma to help you get there. In these cases, the motive could be (and is often) non-dependent

Reply:
We discussed this and agreed that this was not an issue for *induction principles*, but for other lemmas it can be quite useful to have non-dependent versions.